### PR TITLE
8296448: RISC-V: Fix temp usages of heapbase register killed by MacroAssembler::en/decode_klass_not_null

### PR DESCRIPTION
**Pull request by zhengxiaolinX**
Please see the JBS issue for more crash details.

To reproduce using a cross-compiled build:
```
# dump one cds-nocoops.jsa
<java> -XX:-UseCompressedOops -XX:+UseCompressedClassPointers -Xshare:dump -Xlog:cds* -version

# reproduce
<java> -XX:-UseCompressedOops -XX:+UseCompressedClassPointers -Xshare:on -XX:-TieredCompilation \
       -Xlog:cds* -Xlog:gc+metaspace=info -jar renaissance-gpl-0.14.1.jar -r 1 movie-lens
```

`MacroAssembler::en/decode_klass_not_null` uses the heapbase register as a temp register in the interpreter, which may kill the in-use value when enabling C2 compilation and `UseCompressedClassPointers` meanwhile disabling `UseCompressedOops`. C1 won't have this issue for the xheapbase is not its allocation candidate. When CDS is enabled, the narrow klass base is mapped to some address like `0x0000000800000000`, so `MacroAssembler::decode_klass_not_null`, which lacks registers, will use `xheapbase` as a temp to load the klass base and kill the register in the interpreter. So adding a `-XX:+DeoptimizeALot` can speedily reproduce the issue.

To solve this, we shall decouple the xheapbase used as a temp register in `MacroAssembler::en/decode_klass_not_null`. AArch64 has advanced instructions so one register is enough to handle the logic. But in RISC-V we require at least two.

This patch introduces another argument `tmp` to related functions to decouple and eliminate such heapbase usages. One thing that deserves noticing is the `cmp_klass` case, which usually gets used at the beginning of a method entry when `t1` is used as ic holder klass and `t0` is occupied there. These positions are special since nearly all registers are usable except ones used for arguments and special purposes (thread register, etc.). I propose to use a call-clobbered `t2` register here, to keep aligning the `i2c2i_adapter` logic[1].

Tested hotspot tier1\~4 on QEMU; jdk tier1\~tier2 and hotspot tier1\~tier2 on my Hifive unmatched board, and the reproducible movie-lens benchmark.

Thanks,
Xiaolin

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp#L629

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296448](https://bugs.openjdk.org/browse/JDK-8296448): RISC-V: Fix temp usages of heapbase register killed by MacroAssembler::en/decode_klass_not_null

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11010/head:pull/11010` \
`$ git checkout pull/11010`

Update a local copy of the PR: \
`$ git checkout pull/11010` \
`$ git pull https://git.openjdk.org/jdk pull/11010/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11010`

View PR using the GUI difftool: \
`$ git pr show -t 11010`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11010.diff">https://git.openjdk.org/jdk/pull/11010.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp [hunk 1]
```diff
@@ -1406,13 +1406,14 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
 
   const Register ic_reg = t1;
   const Register receiver = j_rarg0;
+  const Register tmp_reg = t2;
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:1409]: No need to introduce 'tmp_reg' here. I think you can use 't2' directly like the change made in file: src/hotspot/cpu/riscv/riscv.ad

File: src/hotspot/cpu/riscv/methodHandles_riscv.cpp [hunk 2]
```diff
@@ -78,7 +78,7 @@ void MethodHandles::verify_klass(MacroAssembler* _masm,
   __ verify_oop(obj);
   __ beqz(obj, L_bad);
   __ push_reg(RegSet::of(temp, temp2), sp);
-  __ load_klass(temp, obj);
+  __ load_klass(temp, obj, temp2);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/methodHandles_riscv.cpp:81]: Could you please help rename 'temp' to 'temp1' at the same time?

File: src/hotspot/cpu/riscv/riscv.ad [hunk 3]
```diff
@@ -1738,7 +1738,7 @@ void MachUEPNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const
   C2_MacroAssembler _masm(&cbuf);
 
   Label skip;
-  __ cmp_klass(j_rarg0, t1, t0, skip);
+  __ cmp_klass(j_rarg0, t1, t0, t2 /* as a tmp */, skip);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv.ad:1741]: You might also want to add comment for 't0' here: /* as a temp */

**Review by RealFYang — COMMENTED** [review]
Thanks for fixing this. Several minor nits.

File: src/hotspot/cpu/riscv/riscv.ad (hunk 3 above)
Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/riscv.ad]: The other two are done, and the comments like this are renamed to a more friendly `/* call-clobbered t2 as a tmp */`.

t0 is always used freely as a tmp register, I guess it is not necessary to add a comment for t0? It seems only the newly-added t2 here matters.

**Review by RealFYang — APPROVED** [review]
Updated change looks good. Thanks for finding and fixing this.

**Comment by yadongw** [comment]
@zhengxiaolinX Nice catch. Looks good, but I'm not sure that all temporary registers are used safely, especially where non-t0 is used?

**Review by yadongw — APPROVED** [review]
lgtm

**Comment by zhengxiaolinX** [comment]
> @zhengxiaolinX Nice catch. Looks good, but I'm not sure that all temporary registers are used safely, especially where non-t0 is used?

Thank you.
I re-examined this patch, and they seem okay to me again.
For the t2 usage in this patch, we may come to the UEP via the `call_stub` when leaving the C++ world, where gcc would help us save alive caller-saved registers. So we can use t2 here. Or, we come to the UEP from the Java world: t2 is SOC so it would have been saved before performing this call if alive.
Some comments in itable stub[1] can also help to make it clear, and in vtable stub[2] it seems we also use a t2 as tmp.
Would this be okay for you :-)?

[1] https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L177-L178
[2] https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L84

**Comment by yadongw** [comment]
> > @zhengxiaolinX Nice catch. Looks good, but I'm not sure that all temporary registers are used safely, especially where non-t0 is used?
> 
> Thank you. I re-examined this patch, and they seem okay to me again. For the t2 usage in this patch, we may come to the UEP via the `call_stub` when leaving the C++ world, where gcc would help us save alive caller-saved registers, so we can use t2 here. Or, we come to the UEP from the Java world: t2 is SOC so it would have been saved before performing this call if alive. Some comments in itable stub[1] can also help to make it clear, and in vtable stub[2] it seems we also use a t2 as tmp. Would this be okay for you :-)?
> 
> [1]
> 
> https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L177-L178
> 
> 
> [2]
> https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L84



> > @zhengxiaolinX Nice catch. Looks good, but I'm not sure that all temporary registers are used safely, especially where non-t0 is used?
> 
> Thank you. I re-examined this patch, and they seem okay to me again. For the t2 usage in this patch, we may come to the UEP via the `call_stub` when leaving the C++ world, where gcc would help us save alive caller-saved registers, so we can use t2 here. Or, we come to the UEP from the Java world: t2 is SOC so it would have been saved before performing this call if alive. Some comments in itable stub[1] can also help to make it clear, and in vtable stub[2] it seems we also use a t2 as tmp. Would this be okay for you :-)?
> 
> [1]
> 
> https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L177-L178
> 
> 
> [2]
> https://github.com/openjdk/jdk/blob/82cbfb5fb0db61f3f1d9f0ceeed20c1cf5474652/src/hotspot/cpu/riscv/vtableStubs_riscv.cpp#L84

It's ok for me, and thanks for the explaination.

**Comment by zhengxiaolinX** [comment]
Thank you for taking time to review this patch! Let's move on then.

/integrate

**Comment by RealFYang** [comment]
I also tested this fix running non-trivial benchmark workloads like Renaissance, SPECjvm2008, SPECjbb2015 with both release & fastdebug builds. So I think it's safe to integrate.
/sponsor

**Comment by andreas-schwab** [comment]
There is a potential conflict with the use of t0 in decode_klass_not_null:

    assert_different_registers(t0, xbase);

since tmp in load_klass defaults to t0.

**Comment by zhengxiaolinX** [comment]
> There is a potential conflict with the use of t0 in decode_klass_not_null:
> 
> ```
> assert_different_registers(t0, xbase);
> ```
> 
> since tmp in load_klass defaults to t0.

Thank you for pointing this out. Yes, code here in `MacroAssembler::decode_klass_not_null` seems to have a potential issue. But in reality I assume it is dead code.

```
// In MacroAssembler::decode_klass_not_null()
//  -> CompressedKlassPointers::base() is non-null also when reaching here
if (CompressedKlassPointers::shift() != 0) {
    assert(LogKlassAlignmentInBytes == CompressedKlassPointers::shift(), "decode alg wrong");
    assert_different_registers(t0, xbase);
    shadd(dst, src, xbase, t0, LogKlassAlignmentInBytes);
}
```
As per Thomas' clear comments here [1], `base > 0 && shift > 0` should not happen. Perhaps we can perform some cleanup here to remove unreachable code that causes confusion. Maybe something like https://github.com/zhengxiaolinX/jdk/commit/973a071e51d116909b9713e52850b4990c025183.

But it is untested due to a lack of resources, so waiting for someone kind to take it over.

BTW the current PR (11010) is still needed because the heapbase register should not be used as a temp register here.

Hope this can explain well.

Best,
Xiaolin


[1] https://github.com/openjdk/jdk/blob/4dcc7f3f2629e857b20f72e99189db8781aa65ff/src/hotspot/share/oops/compressedKlass.cpp#L103-L104

**Comment by andreas-schwab** [comment]
Thanks for your explanation.

This came up while trying to backport the RISC-V port to jdk16 (for
bootstrapping purpose).  In jdk16 base != 0 and shift != 0 is possible,
apparently due to lack of JDK-8265705.

**Comment by zifeihan** [comment]
> > There is a potential conflict with the use of t0 in decode_klass_not_null:
> > ```
> > assert_different_registers(t0, xbase);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > since tmp in load_klass defaults to t0.
> 
> Thank you for pointing this out. Yes, code here in `MacroAssembler::decode_klass_not_null` seems to have a potential issue. But in reality I assume it is dead code.
> 
> ```
> // In MacroAssembler::decode_klass_not_null()
> //  -> CompressedKlassPointers::base() is non-null also when reaching here
> if (CompressedKlassPointers::shift() != 0) {
>     assert(LogKlassAlignmentInBytes == CompressedKlassPointers::shift(), "decode alg wrong");
>     assert_different_registers(t0, xbase);
>     shadd(dst, src, xbase, t0, LogKlassAlignmentInBytes);
> }
> ```
> 
> As per Thomas' clear comments here [1], `base > 0 && shift > 0` should not happen. Perhaps we can perform some cleanup here to remove unreachable code that causes confusion. Maybe something like [zhengxiaolinX@973a071](https://github.com/zhengxiaolinX/jdk/commit/973a071e51d116909b9713e52850b4990c025183).
> 
> But it is untested due to a lack of resources, so waiting for someone kind to take it over.
> 
> BTW the current PR (11010) is still needed because the heapbase register should not be used as a temp register here.
> 
> Hope this can explain well.
> 
> Best, Xiaolin
> 
> [1]
> 
> https://github.com/openjdk/jdk/blob/4dcc7f3f2629e857b20f72e99189db8781aa65ff/src/hotspot/share/oops/compressedKlass.cpp#L103-L104

@zhengxiaolinX Hi, I have tested tier1(fastdebug) with this patch [zhengxiaolinX@973a071](https://github.com/zhengxiaolinX/jdk/commit/973a071e51d116909b9713e52850b4990c025183) and everything works fine.

**Pull request closed without merging**